### DISCUSSION TITLE
Userscript to show links to the bridged profiles of Bluesky users who have already opted in

### DISCUSSION THREAD

**Discussion by afontenot** [snarfed/bridgy-fed#1456] [Show and tell]
I don't have any desire for random DMs to be sent on my behalf to Bluesky users who don't know me, and most users appear to have DMs semi-closed anyway. Instead, I threw together a userscript that shows a link on the profile page if the user is bridged. Probably a little flaky, sorry. If someone wanted to extend it to show a link or icon on other pages than the profile page, that would be cool.

https://gist.github.com/afontenot/63722433f34b799c7e9c9aa4f07b34e2

Example:

![image](https://github.com/user-attachments/assets/a0188a95-c49a-4ce2-aaa0-fa530ff4db66)

**Comment by afontenot** [comment]
Sadly I wouldn't recommend browsing around random profiles looking for people bridged, only a *very* small fraction of users have opted in at this point.

**Comment by Tamschi** [comment]
Related: #1388

Checking the followers is unfortunately not going to be entirely accurate, since the opt-out and Chat commands make that not map 1:1 to who's bridged currently. Checking _whom the bridge follows_ would be pretty accurate though!

**Reply by afontenot** [reply]
I went with the approach I did because I noticed that the bridge user is following fewer users than are following it, and wasn't sure of the reason for this. Is the gap between the two (1.4k or ~8%) users who have followed the bridge but *also* explicitly opted out?

**Reply by afontenot** [reply]
Okay, also I've had a look at switching to the follows API and there's a pretty serious issue, in that the API appears not to actually show all the follows. Edit: one issue here was that I wasn't handling `handle.invalid` from the API. After fixing that, I'm seeing 15295 handles out of 16.3k.

Take a look at [this API request for follows](https://public.api.bsky.app/xrpc/app.bsky.graph.getFollows?actor=did:plc:xbifsywyv5pka5jlknhv5yv3&limit=100) versus [this API request for followers](https://public.api.bsky.app/xrpc/app.bsky.graph.getFollowers?actor=did:plc:xbifsywyv5pka5jlknhv5yv3&limit=100). Despite not having a cursor, the "follows" request only shows 99 follows for me, and the "followers" request shows 100 as expected. Using the cursor to go further down the list quickly shows worse issues.

The [API docs](https://docs.bsky.app/docs/api/app-bsky-graph-get-follows) don't mention anything about filtering happening on this endpoint, so I can't imagine what this could be.

As an example, after getting all the handles returned by the API, `handles.has("samuel.bsky.team")` returns `false`, despite this user being bridged.

**Reply by Tamschi** [reply]
> I went with the approach I did because I noticed that the bridge user is following fewer users than are following it, and wasn't sure of the reason for this. Is the gap between the two (1.4k or ~8%) users who have followed the bridge but _also_ explicitly opted out?

A combination of that, glitches where the bridge missed new followers and [unfortunately also sporadic outbound reliability issues towards Bluesky](https://github.com/snarfed/arroba/issues/34) until very recently (which goes in either direction). Still, the set of follows should be most accurate. Bluesky accounts can also unfollow the bridge and still remain bridged, but the bridge should have unfollowed them.

> Okay, also I've had a look at switching to the follows API and there's a pretty serious issue, in that the API appears not to actually show all the follows. Edit: one issue here was that I wasn't handling `handle.invalid` from the API. After fixing that, I'm seeing 15295 handles out of 16.3k.
> 
> […]

That seems about right. It would be nice if Bluesky made the account's DID available as `data-`attribute on relevant DOM elements though, so that browser extensions could more easily track accounts reliably and handle those with invalid handles too.

**Comment by snarfed** [comment]
Fun! Love it!

**Comment by afontenot** [comment]
Update, in case anyone is subscribed to this discussion: after some frustration with the follow / following APIs, I changed and simplified the userscript. It no longer caches all the handles as before, and instead sends 2 API requests on each profile page load - one to resolve the handle (unfortunately unavoidable?), and one to the `app.bsky.graph.getRelationships` endpoint to get the follow / following relationship between the bridge actor and the user.

I'm showing the link if *either* one follows the other. This is unfortunate but there are just way too many false negatives - about 9/10 users I check who I know are bridged are missing from the `getFollows` graph. The only downside of a false positive is that you will click on a link and get a 404 on your Fediverse instance because the user isn't actually bridged.

**Reply by afontenot** [reply]
Fixed an issue where I accidentally removed the detection code for same-document navigations, sorry about that!
